internal/controller/factory/statefulset.go

```diff
@@ -482,3 +374,269 @@ func mergeWithDefaultProbe(probe *corev1.Probe, defaultProbe corev1.Probe) *core
+	}
+
+	listenClientURL := fmt.Sprintf("%s://0.0.0.0:2379", clientProtocol)
+	advertiseClientURL := fmt.Sprintf("%s://$(POD_NAME).%s.$(POD_NAMESPACE).svc:2379",
+		clientProtocol, cluster.Name)
+	initialAdvertisePeerURL := fmt.Sprintf("https://$(POD_NAME).%s.$(POD_NAMESPACE).svc:2380", cluster.Name)
+
+	args = append(args, "--listen-client-urls="+listenClientURL)
+	args = append(args, "--advertise-client-urls="+advertiseClientURL)
+	args = append(args, "--initial-advertise-peer-urls="+initialAdvertisePeerURL)
+
+	return args
+}
+
+func etcdExtraFlags(options map[string]string) []string {
+	args := []string{}
+
+	for name, value := range options {
+		flag := "--" + name
+		if len(value) == 0 {
+			args = append(args, flag)
+		} else {
+			args = append(args, fmt.Sprintf("%s=%s", flag, value))
+		}
+	}
+	return args
+}
+
+func generateSecretVolumes(cluster *etcdaenixiov1alpha1.EtcdCluster) map[string]corev1.Volume {
+	volumesMap := make(map[string]corev1.Volume)
+
+	addSecretVolume := func(name, secretName string) {
+		volumesMap[name] = createSecretVolume(name, secretName)
+	}
+
+	if cluster.Spec.Security != nil {
+		if cluster.Spec.Security.TLS.PeerSecret != "" {
+			addSecretVolume(peerTrustedCACertificate.Name(), cluster.Spec.Security.TLS.PeerTrustedCASecret)
+			addSecretVolume(peerCertificate.Name(), cluster.Spec.Security.TLS.PeerSecret)
+		}
+		if cluster.Spec.Security.TLS.ServerSecret != "" {
+			addSecretVolume(serverCertificate.Name(), cluster.Spec.Security.TLS.ServerSecret)
+		}
+		if cluster.Spec.Security.TLS.ClientSecret != "" {
+			addSecretVolume(clientTrustedCACertificate.Name(), cluster.Spec.Security.TLS.ClientTrustedCASecret)
+		}
+	}
+
+	return volumesMap
+}
+
+func createSecretVolume(name, secretName string) corev1.Volume {
+	return corev1.Volume{
+		Name: name,
+		VolumeSource: corev1.VolumeSource{
+			Secret: &corev1.SecretVolumeSource{
+				SecretName: secretName,
+			},
+		},
+	}
+}
+
+func generateDataVolume(cluster *etcdaenixiov1alpha1.EtcdCluster) corev1.Volume {
+	var dataVolumeSource corev1.VolumeSource
+
+	if cluster.Spec.Storage.EmptyDir != nil {
+		dataVolumeSource = corev1.VolumeSource{EmptyDir: cluster.Spec.Storage.EmptyDir}
+	} else {
+		dataVolumeSource = corev1.VolumeSource{
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: GetPVCName(cluster),
+			},
+		}
+	}
+
+	return corev1.Volume{
+		Name:         etcdConfig.VolumeName(),
+		VolumeSource: dataVolumeSource,
+	}
+}
+
+func updateOrAddDataVolumeMount(volumeMounts []corev1.VolumeMount,
+	volumeName, mountPath string, readOnly bool) []corev1.VolumeMount {
+	found := false
+
+	for i := range volumeMounts {
+		if volumeMounts[i].Name == volumeName {
+			volumeMounts[i].ReadOnly = readOnly
+			volumeMounts[i].MountPath = mountPath
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      volumeName,
+			ReadOnly:  readOnly,
+			MountPath: mountPath,
+		})
+	}
```

Comment:
I'd suggest to use `slices`:

```go
if idx := slices.IndexFunc(volumeMounts, func(...) bool {...}; idx < 0 {
    volumeMounts = append(...)
}
```

Reply:
A lot of this custom merging logic could be handled nicely with `StrategicMergePatch`.